compiler/rustc_builtin_macros/src/eii.rs

```diff
@@ -0,0 +1,414 @@
+                    kind: ast::TyKind::Tup(ThinVec::new()),
+                    span,
+                    tokens: None,
+                }),
+                expr: Some(Box::new(ast::Expr {
+                    id: DUMMY_NODE_ID,
+                    kind: ast::ExprKind::Block(
+                        Box::new(ast::Block {
+                            stmts: thin_vec![ast::Stmt {
+                                id: DUMMY_NODE_ID,
+                                kind: ast::StmtKind::Item(Box::new(ast::Item {
+                                    attrs: thin_vec![], // FIXME: re-add some original attrs
+                                    id: DUMMY_NODE_ID,
+                                    span: item_span,
+                                    vis: ast::Visibility {
+                                        span,
+                                        kind: ast::VisibilityKind::Inherited,
+                                        tokens: None
+                                    },
+                                    kind: ItemKind::Fn(default_func),
+                                    tokens: None,
+                                })),
+                                span
+                            }],
+                            id: DUMMY_NODE_ID,
+                            rules: ast::BlockCheckMode::Default,
+                            span,
+                            tokens: None,
+                        }),
+                        None,
+                    ),
+                    span,
+                    attrs: ThinVec::new(),
+                    tokens: None,
+                })),
+                define_opaque: None,
+            })),
+            tokens: None,
+        }))
+    }
+
+    let decl_span = span.to(func.sig.span);
+
+    let abi = match func.sig.header.ext {
+        // extern "X" fn  =>  extern "X" {}
+        ast::Extern::Explicit(lit, _) => Some(lit),
+        // extern fn  =>  extern {}
+        ast::Extern::Implicit(_) => None,
+        // fn  =>  extern "Rust" {}
+        ast::Extern::None => Some(ast::StrLit {
+            symbol: sym::Rust,
+            suffix: None,
+            symbol_unescaped: sym::Rust,
+            style: ast::StrStyle::Cooked,
+            span,
+        }),
+    };
+
+    // ABI has been moved to the extern {} block, so we remove it from the fn item.
+    func.sig.header.ext = ast::Extern::None;
+
+    // And mark safe functions explicitly as `safe fn`.
+    if func.sig.header.safety == ast::Safety::Default {
+        func.sig.header.safety = ast::Safety::Safe(func.sig.span);
+    }
+
+    // extern "…" { safe fn item(); }
+    let extern_block = Box::new(ast::Item {
+        attrs: ast::AttrVec::default(),
+        id: ast::DUMMY_NODE_ID,
+        span,
+        vis: ast::Visibility { span, kind: ast::VisibilityKind::Inherited, tokens: None },
+        kind: ast::ItemKind::ForeignMod(ast::ForeignMod {
+            extern_span: span,
```

Comment:
Could be computed from the fn's extern span, everything but the generated Rust has one.

Various other uses of the fallback span may also have better spans to be used